go/pkg/client/tree.go

```diff
@@ -165,21 +165,85 @@ func getExecRootRelPaths(absPath, execRoot, workingDir, remoteWorkingDir string)
 	return relPath, remoteRelPath, nil
 }
 
+// evalParentSymlinks replaces each parent element in relPath with its target if it's a symlink.
+//
+// Returns the resolved absolute path with a list of parent symlinks (absolute paths) if any.
+// The basename of relPath is not resolved. It remains a symlink if it is one.
+// There are no guaranteed relationship between relPath and the returned path beyond having the same basename.
+// The returned paths may not be clean.
+// Any errors would be from accessing files.
+// Example: execRoot=/a relPath=b/c/d/e.go, b->bb, evaledPath=/a/bb/c/d/e.go, symlinks=[/a/b]
+func evalParentSymlinks(execRoot, relPath string, fmdCache filemetadata.Cache) (evaledPath string, symlinks []string, err error) {
+	b := strings.Builder{}
+	b.WriteString(execRoot)
+	ps := strings.Split(relPath, string(filepath.Separator))
+	lastIndex := len(ps) - 1
```

Comment:
Three reasons:
- it would not make use of the file metadata cache which means more IO, 
- it would double processing for positive cases,
- and it would resolve the basename if it was a symlink which is not desired.

Reply:
- it would not make use of the file metadata cache which means more IO,
We could cache the symlink evaluation for the parent directory though, no?
- it would double processing for positive cases,
Yes, but my suspicion is that cases with symlinks are significantly less than cases without, and doing the string processing here for every input may be too high of a cost. At the very least we should make sure this doesn't regress performance in some sample builds?
- and it would resolve the basename if it was a symlink which is not desired.
Why is that not desired? This is just to check that we need extra symlink processing, which again should be rare compared to physical files.

Reply:
If the file itself is a symlink, `EvalSymlink(path) != path`, but `EvalSymlink(Dir(path)) == path` may be true.

Caching the output of `EvalSymlinks(Dir(path))` is useful when many files have the same direct parent. Otherwise, the cache needs to include ancestors as well. Even then, every path needs to be traversed backwards to do cache lookups, while cache writes need a forward traversal to properly expand targets.

I'm only aware of one build that requires symlinks with no baseline for regression detection.

Reply:
Thanks for clarifying, I meant `EvalSymlink(Dir(path))`. What I'm thinking is this cache will just gate the logic here, but you still need to do the splitting and checking of ancestors like you do now.

> I'm only aware of one build that requires symlinks with no baseline for regression detection.
Ack, I think its ok as is then and let's optimize when we use it in practice.